assist: support recording non-interactive forwarded sessions

hugoShaka opened the pull request:
Fixes https://github.com/gravitational/teleport.e/issues/1773

There were two issues blocking openssh and proxy-recorded sessions from being captured:
- the forwarder did not save the session information in the environment variables. The forwarder only passed variables like TELEPORT_SESSION and SSH_TELEPORT_RECORD_NON_INTERACTIVE without adding them into the context. Because of this, the execution handler followed the regular path, not the Assist one. Saving `TELEPORT_*` and `SSH_TELEPORT_*` variables into the context fixes the issue. 
- The writer used to record the session was set for local executions, but not for remote executions. This caused the recordings to be empty.

With this PR, proxy-recording and openssh Assist sessions are now recorded and displayed in assist after a page refresh.

The PR introduces Assist integration tests and fixes two non-related tests:
- openssh node join was not waiting properly the SSH node to come up
- openAI mock server was not serving `/v1` routes, which caused a few `lib/web` tests to throw errors

hugoShaka committed:
assist: support recording non-interactive forwarded sessions

jakule commented on lib/srv/forward/sshserver.go:
```diff
@@ -1366,6 +1373,15 @@ func (s *Server) handleEnvs(ctx context.Context, ch ssh.Channel, req *ssh.Reques
+	// We want to save all environment variables even if the ssh server rejects
+	// them, as we still need their information (e.g. the session ID).
+	for k, v := range envs {
+		scx.SetEnv(k, v)
```
Will this send the environment variables to the sshd? I think this is a big change in terms of functionality and potential performance (setting environment variables in SSH is slow).

hugoShaka replied:
The environment variables are already sent to sshd, we just keep them in memory as well. I see two risks with this change:
- if a user sets a lot of XXL environment variables, this will consume more memory
- users might set those environment variables themselves, there might be a security risk if a user changes its TELEPORT_SESSION var. This PR doesn't introduce this risk: it was already here since the Assist non-interactive recording logic.

zmb3 reviewed (commented):
No tests?

jakule commented on lib/client/session.go:
```diff
@@ -236,6 +236,7 @@ func (ns *NodeSession) createServerSession(ctx context.Context) (*tracessh.Sessi
 		}
 	}
 	// pass environment variables set by client
+	log.Tracef("Propagating all environment variables")
```
I don't think this log is necessary.

hugoShaka replied:
```suggestion
```

hugoShaka committed:
assist: add integration tests for assist command recording on agentless

This also fixes a bugged openssh integration test check.

hugoShaka committed:
Mock OpenAI API in integration tests.

Also fixes the OpenAI mock handler that was not supporting the "/v1/*"
routes used in some web/ tests.

hugoShaka commented:
> No tests?

I was hoping not to get this comment 😅 . This was really painful to test, but after 48 hours, I got an integration test. We'll very likely expand the Asisst testing, hence the dedicated test directory.

When writing the test, I stumbled across a bug in the OpenSSH joining test: the label matching was broken, but the test was green because the Teleport process also started its own SSH node and always considered this a success (it was not). The PR fixes this.

I also added the ability to configure a Teleport-wide OpenAI client (to mock in various tests) and fixed a small regression in the OpenAI mock handler (it was rejecting calls prefixed with "/v1", but the `lib/web` test suite adds this prefix).

hugoShaka committed:
Docstrings + adressing again a feedback that got lost during a rebase

hugoShaka committed:
restore signer function signature

hugoShaka commented on api/utils/sshutils/test.go:
```diff
@@ -25,8 +25,11 @@ import (
 	"golang.org/x/crypto/ssh"
 
 	"github.com/gravitational/teleport/api/constants"
+	"github.com/gravitational/teleport/lib/defaults"
 )
 
+const defaultPrincipal = defaults.Localhost
```
Oooops

hugoShaka commented on integration/assist/command_test.go:
```diff
@@ -0,0 +1,451 @@
+	testClusterName   = "teleport.example.com"
+)
+
+// TestIntegrationCRUD starts a Teleport cluster and using its Proxy Web server,
+// tests the CRUD operations over the Integration resource.
```
Update docstring

hugoShaka commented on integration/assist/command_test.go:
```diff
@@ -0,0 +1,451 @@
+	require.NoError(t, auth.UpsertPassword(testUser, []byte(userPassword)))
+
+	// Get user certs and build a client
+	userKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	require.NoError(t, err)
+	userPubKey, err := ssh.NewPublicKey(&userKey.PublicKey)
+	require.NoError(t, err)
+	testCertsReq := libauth.GenerateUserTestCertsRequest{
+		Key:            ssh.MarshalAuthorizedKey(userPubKey),
+		Username:       user.GetName(),
+		TTL:            time.Hour,
+		Compatibility:  constants.CertificateFormatStandard,
+		RouteToCluster: testClusterName,
+	}
```
TODO: put this in `GenerateUserTestCerts`

hugoShaka commented on integration/assist/command_test.go:
```diff
@@ -0,0 +1,451 @@
+// starts a mock SSH server.
+func registerAndSetupMockSSHNode(t *testing.T, ctx context.Context, testDir string, rc *helpers.TeleInstance) types.Server {
+	// Reserve the listener for the SSH server. We can't start the SSH server right now because we need to get a valid certificate.
+	// The certificate needs proper principals, whichj implies knowing the node ID. This only happens after the node has joined.
```
```suggestion
	// The certificate needs proper principals, which implies knowing the node ID. This only happens after the node has joined.
```

jakule reviewed (approved):
Approved with a few comments

hugoShaka commented on integration/assist/command_test.go:
```diff
@@ -0,0 +1,451 @@
+	require.Equal(t, "stdout", perNodeEnvelope.MsgType)
+	output, err := base64.StdEncoding.DecodeString(perNodeEnvelope.Payload)
+	require.NoError(t, err)
+	require.Equal(t, testCommandOutput, string(output))
```
I think this is covered by `require.Equal(t, "stdout", perNodeEnvelope.MsgType)`, when an execution error happens, the value is `teleport-error`. I'll add a comment explaining the line

hugoShaka committed:
Address Jakub's feedback + diverse improvements

- improve comments
- simpler file creation
- more efficient rsa key generation
- use assert instead of require in other goroutines
- only save env vars if they are Teleport-related
- move more logic into newTestCredentials for readability

zmb3 commented on integration/assist/command_test.go:
```diff
@@ -0,0 +1,458 @@
+	// "pty-req", "env" and "exec". Here we don't output anything and start a
+	// routine consuming requests and waiting for the "exec" one.
+	go func(in <-chan *ssh.Request) {
+		for req := range in {
```
What's going to close the `requests` channel so this goroutine eventually terminates?

hugoShaka committed:
Update lib/srv/forward/sshserver.go

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

hugoShaka committed:
Add a timeout in the ssh handler coroutine

hugoShaka committed:
Merge branch 'master' into hugo/fix-e1773-assist-proxy-side-recording
